include/mbgl/shaders/mtl/hillshade.hpp

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <mbgl/shaders/hillshade_layer_ubo.hpp>
+#include <mbgl/shaders/shader_source.hpp>
+#include <mbgl/shaders/mtl/common.hpp>
+#include <mbgl/shaders/mtl/shader_program.hpp>
+
+namespace mbgl {
+namespace shaders {
+
+template <>
+struct ShaderSource<BuiltIn::HillshadeShader, gfx::Backend::Type::Metal> {
+    static constexpr auto name = "HillshadeShader";
+    static constexpr auto vertexMainFunction = "vertexMain";
+    static constexpr auto fragmentMainFunction = "fragmentMain";
+
+    static constexpr AttributeInfo attributes[] = {
+        {0, gfx::AttributeDataType::Short2, 1, "a_pos"},
+        {1, gfx::AttributeDataType::Short2, 1, "a_texture_pos"},
+    };
+    static constexpr UniformBlockInfo uniforms[] = {
+        MLN_MTL_UNIFORM_BLOCK(2, true, true, HillshadeDrawableUBO),
+        MLN_MTL_UNIFORM_BLOCK(3, false, true, HillshadeEvaluatedPropsUBO),
+    };
+    static constexpr TextureInfo textures[] = {
+        {0, "u_image"},
+    };
+
+    static constexpr auto source = R"(
+
+struct VertexStage {
+    short2 pos [[attribute(0)]];
+    short2 texture_pos [[attribute(1)]];
+};
+
+struct FragmentStage {
+    float4 position [[position, invariant]];
+    float2 pos;
+};
+
+struct alignas(16) HillshadeDrawableUBO {
+    float4x4 matrix;
+    float2 latrange;
+    float2 light;
+};
+
+struct alignas(16) HillshadeEvaluatedPropsUBO {
+    float4 highlight;
+    float4 shadow;
+    float4 accent;
+};
+
+FragmentStage vertex vertexMain(thread const VertexStage vertx [[stage_in]],
+                                device const HillshadeDrawableUBO& drawable [[buffer(2)]]) {
+    
+    const float4 position = drawable.matrix * float4(float2(vertx.pos), 0, 1);
+    float2 pos = float2(vertx.texture_pos) / 8192.0;
+    pos.y = 1.0 - pos.y;
+    
+    return {
+        .position    = position,
+        .pos         = pos,
+    };
+}
+
+half4 fragment fragmentMain(FragmentStage in [[stage_in]],
+                            device const HillshadeDrawableUBO& drawable [[buffer(2)]],
+                            device const HillshadeEvaluatedPropsUBO& props [[buffer(3)]],
+                            texture2d<float, access::sample> image [[texture(0)]],
+                            sampler image_sampler [[sampler(0)]]) {
+
+    /*if (permutation.overdrawInspector) {
+        return half4(1.0);
+    }*/
+
+    float4 pixel = image.sample(image_sampler, in.pos);
+
+    float2 deriv = ((pixel.rg * 2.0) - 1.0);
+
+    // We divide the slope by a scale factor based on the cosin of the pixel's approximate latitude
+    // to account for mercator projection distortion. see #4807 for details
```

Comment:
I know this comment is also in the GL version, but I am not sure what it refers to.

Reply:
I didn't focus to understand the shaders, just converted them from GL to Metal.